logging: syst: add support for catalog messages

dcpleung opened the pull request:
MIPI Sys-T catalog messages are similar to dictionary logging where an ID is emitted instead of the whole format string to minimize log data. This adds the necessary bits to determine when to emit catalog messages.

This also changes how the collateral file for dictionary logging is created.

See individual commits for details.

dcpleung committed:
scripts: logging/dictionary: extract actual strings for db

This changes the database generation to actually extracting
individual strings instead of stuffing the whole binary sections
into the database. This allows the generation script to be
extended to accommodate more output formats.

Note that if CONFIG_LOG2_FMT_SECTION is enabled, the format
strings are in log_strings_sections, and also have associated
debug symbols in DWARF. So there is no need to manually
extract them.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: logging/dictionary: fixes pylint warnings

Fixes some easy pylint warnings.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: logging/dictionary: can specify output format

This changes the script to allow output format to be specified.
Currently, only JSON is support. This will allow supporting
other formats in the future.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: logging/dictionary: update test for combined strings

Compilers often combine strings to conserve space, if one string is
a perfect substring of another one towards the end. So add another
string in the test to make sure dictionary logging is still working
correctly under this scenario.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: logging/dictionary: generate MIPI Sys-T collateral

This adds the ability to the dictionary logging database
generation script to output MIPI Sys-T collateral in XML.
This will allow usage of catalog logging under Sys-T, which
is similar to dictionary logging.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: syst: add support for catalog messages

MIPI Sys-T catalog messages are similar to dictionary logging
where an ID is emitted instead of the whole format string to
minimize log data. This adds the necessary bits to determine
when to emit catalog messages.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
west.yml: update mipi-sys-t for catalog msg support

This updates the mipi-sys-t module to support prepared argument
list for catalog messages.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: syst: v1/imm: remove temp buffer for sync string output

The MIPI Sys-T library can now take a format string with a variable
argument list so there is no need for the temporary buffer anymore.

This saves some stack space in v1 immediate mode.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
samples: logging/syst: update sample for combined strings

Compilers often combine strings to conserve space, if one string is
a perfect substring of another one towards the end. So add another
string in the test to make sure sys-t with catalog messages is still
working correctly under this scenario.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
samples: syst: allow some 64-bit boards

This adds qemu_x86_64 and qemu_cortex_a53 to make sure 64-bit
support for Sys-T is not broken in the future.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andyross commented on include/linker/utils.h:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @return True if address identified within read only section.
+ */
+static inline bool lnkr_is_in_rodata(const void *addr)
```
Might want to limit the scope here to a "is_string_literal()" predicate, since AFAICT that's what it's being used to detect.  There are platforms where it's convenient to link data directly into .text, where there are multiple regions for immutable data (the DSP is moving in this direction with __imr) etc... The toolchain behavior with string literals isn't well specified or portable either, but it's likely to be more easily pinned down in the face of weird platform linkage.

dcpleung replied:
This corresponds to the python script to extract strings from ELF file, where strings are extracted from inside `.rodata`.

Note that this will also be used for cbprintf once the bug on SPARC is fixed there (https://github.com/zephyrproject-rtos/zephyr/issues/42208).

nordic-krch replied:
Done PR for cbprintf #42577

dcpleung replied:
Thank you. Will rebase.

nordic-krch replied:
@dcpleung Could you extract this one into separate PR? That could get in earlier. While working on cbprintf extension i could use it there.
Also i would not abbreviate `linker` to `lnkr`. Function name is not that long.

nordic-krch replied:
Regarding @andyross comment, maybe you could add note to the description that variable may be in read only section/memory even though function return `false`.

dcpleung replied:
Submitted the linker commit @ https://github.com/zephyrproject-rtos/zephyr/pull/43168

andyross reviewed (dismissed):
One nitpick on the linkage stuff.  Everything else looks great to me, but then I'm not a logging expert so apply salt.

dcpleung committed:
samples: logging/syst: build catalog samples

This adds to sample.yaml to build for MIPI Sys-T catalog message
support.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Failure in bluetooth tests is due to https://github.com/zephyrproject-rtos/zephyr/issues/42608

tejlmand commented on CMakeLists.txt:
```diff
@@ -1656,7 +1656,11 @@ list(APPEND
 if(CONFIG_LOG_DICTIONARY_SUPPORT)
   set(LOG_DICT_DB_OUTPUT --json=${PROJECT_BINARY_DIR}/log_dictionary.json)
+elseif(CONFIG_LOG_MIPI_SYST_USE_CATALOG)
+  set(LOG_DICT_DB_OUTPUT --syst=${PROJECT_BINARY_DIR}/mipi_syst_collateral.xml)
```
please lower case to indicate this is locally used variable.
```suggestion
  set(log_dict_db_output --syst=${PROJECT_BINARY_DIR}/mipi_syst_collateral.xml)
```

tejlmand reviewed (commented):
Thanks, generally good looking.

A couple of minor nits.

dcpleung commented on scripts/logging/dictionary/database_gen.py:
```diff
@@ -521,9 +544,18 @@ def main():
+        rawstr_permutations = True
+
+    if not has_output:
         logger.error("Need to specify output file.")
```
I used `argparse.add_mutually_exclusive_group()` where either `--json` or `--syst` must be specified. So there is no need for this error anymore.

dcpleung committed:
logging: support runtime packaging without format strings in mem

This adds support for runtime packaging with tagged arguments
so that cbvprintf_package() can function without the format
strings in memory. This allows the format strings to be removed
from final binary to converse space.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
lib: os: cbprint: enable tagged arguments for packaging

This adds some bits to support tagged arguments to be used for
packaging. If enabled, the packaging function no longer looks at
the format strings to determine the types of arguments, but
instead, each argument is tagged with a type by preceding it
with another argument as type (integer). This allows the format
strings to be removed from the final binary to conserve space.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: add pass-thru package flags for runtime packaging

This adds a field to z_log_msg2_runtime_create() to pass
through packaging flags to cbprintf_package().

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: logging/log_msg2: add a missing test to suite

The test test_mode_size_str_with_strings() is defined but was not
added to the test suite, so it never ran before. Add this to
the suite so this can be tested.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: logging/log_msg2: add a few bits for tagged arguments

This adds a few bits to support testing for runtime tagged arguments
when CONFIG_LOG2_RUNTIME_USE_TAGGED_ARGUMENTS is enabled.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: logging/log_api: add a few bits for tagged arguments

This adds a few bits to support testing for runtime tagged arguments
when CONFIG_LOG2_RUNTIME_USE_TAGGED_ARGUMENTS is enabled.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
This won't work in C++... so need to start over.
